File: 01_Array/Q10_45. Jump Game II/solution.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        total_jumps, coverage, last_jump_idx = 0,0,0
        destination = n - 1

        if n == 1: return 0

        for i in range(0, n):
            coverage = max(coverage, i + nums[i])

            if i == last_jump_idx:
                last_jump_idx = coverage
                total_jumps += 1

                if coverage >= destination: return total_jumps

        return total_jumps



# Time Complexity -> O(n) 
# Space C0mplexity -> O(1)

File: 01_Array/Q10_45. Jump Game II/solution.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int totalJumps = 0;
        int destination = nums.size() - 1;
        int coverage = 0;
        int lastJumpIdx = 0;

        if(nums.size() == 1) return 0;

        for(int i = 0; i < nums.size(); i++){
            coverage = max(coverage, i + nums[i]);

            if(i == lastJumpIdx){
                lastJumpIdx = coverage;
                totalJumps++;

                if(coverage >= destination) return totalJumps;
            }
        }
        return totalJumps;
    }
};



// Time Complexity ->O(n)
// Space Complexity ->O(1)
